src/System.Private.Xml/src/System/Xml/Xsl/Xslt/XPathPatternBuilder.cs

```diff
@@ -328,7 +328,7 @@ private QilNode UnexpectedToken(string tokenName)
         {
             string prompt = string.Format(CultureInfo.InvariantCulture, "Internal Error: {0} is not allowed in XSLT pattern outside of predicate.", tokenName);
             Debug.Assert(false, prompt);
-            throw new Exception(prompt);
```

Comment:
this exception after Assert is strange. Maybe throw exception on Number() and Variable() methods

Reply:
The reason it's doing a `throw` is so that the compiler recognizes the fact that there's no method return, which is required because `Number` and `Variable` require a return....
The big change here would be removing the assert, not quite so much changing the exception.

Returning the parameter name is going to be mysterious here, since `tokenName` is from a private method.  I'm also unsure if `ArgumentException` is the correct type here, as opposed to something like `NotSupportedException`.

Reply:
@Clockwork-Muse i agree. Maybe removing assert and throw NotSupportedException() is better.